std-bits/base/src/main/java/org/enso/base/FileLineReader.java

```diff
@@ -0,0 +1,370 @@
+
+            if (index % 100000 == 0) {
+              LOGGER.log(Level.INFO, "Scanned Lines: {0}", index);
+            }
+            index++;
+
+            // If no filter we can record the start of the next line.
+            if (filter == null && index == rowMap.getSize()) {
+              rowMap.add(buffer.position() + position);
+            }
+          }
+
+          // Fast-forward if needed
+          if (filter != null && index < rowMap.getSize()) {
+            int newPosition = Math.min(bufferSize, (int) (rowMap.get(index) - position));
+            buffer.position(newPosition);
+          }
+        } else {
+          // Read more if we need to
+          if (!buffer.hasRemaining()) {
+            position = linePosition;
+            bufferSize = (int) Math.min(BUFFER_SIZE, (length - position));
+            truncated = bufferSize != (length - position);
+            buffer = channel.map(FileChannel.MapMode.READ_ONLY, position, bufferSize);
+          }
+        }
+      }
+
+      if (!truncated && !buffer.hasRemaining() && rowMap.get(rowMap.getSize() - 1) != length) {
+        // Add the last line to mark reached the end.
+        rowMap.add(length);
+      }
+
+      return output;
+    }
+  }
+
+  // ** Scans forward in a file reading line by line until it finds a line that matches the new
+  // filter. */
+  public static long findFirstNewFilter(
+      File file,
+      long length,
+      LongArrayList rowMap,
+      int endAt,
+      Charset charset,
+      Function<ByteArrayOutputStreamWithContains, String> filter,
+      Function<ByteArrayOutputStreamWithContains, String> newFilter)
+      throws IOException {
+    final CancellationToken token = new CancellationToken();
+    final List<Long> result = new ArrayList<>();
+    BiConsumer<Integer, String> action =
+        (index, line) -> {
+          var bytes = line.getBytes(charset);
+          var outputStream = ByteArrayOutputStreamWithContains.fromByteArray(bytes);
+          if (newFilter.apply(outputStream) != null) {
+            result.add(rowMap.get(index));
+            token.cancel();
+          }
+        };
+    innerForEachLine(file, length, rowMap, 0, endAt, charset, filter, action, token);
+    return result.isEmpty() ? rowMap.get(rowMap.getSize() - 1) : result.get(0);
+  }
+
+  // ** Creates a filter that checks if the line contains the given string. */
+  public static Function<ByteArrayOutputStreamWithContains, String> createContainsFilter(
+      String contains, Charset charset) {
+    if (isUnicodeCharset(charset)) {
+      var nfcVersion = Normalizer2.getNFCInstance().normalize(contains);
+      var nfdVersion = Normalizer2.getNFDInstance().normalize(contains);
+      if (!nfcVersion.equals(nfdVersion)) {
+        // Need to use Unicode normalization for equality.
+        return (outputStream) -> {
+          var line = outputStream.toString(charset);
+          return Core_Text_Utils.compare_normalized(contains, line) == 0 ? line : null;
```

Comment:
Is this surely right?

The byte version does `contains`, but here we do `compare_normalized` which is an `==` equality comparison. So this will match only if the line matches exactly the `contains` string, not if it merely contains it. That does not seem right, does it?

Instead we should use `Text_Utils.contains` which performs Unicode-aware string search.